Convert service to use ancestry

kbrock opened the pull request:
This is the main version of a fix for 5.4. Instead of using recursive calls, we are using ancestry.
The original code cut down the the view services screen from **68 seconds to 10**, and number of queries from **3.2k to 860**.

This uses ancestry for the calls.

/cc @Fryguy Think this is pretty straight forward
/cc @bzwei fyi
/cc @gmcculloug think you had a bunch of insight into this one.

kbrock commented:
@bzwei  looks like service and service_templates are passed around here.
Looks like I need to add `ancestry` to service templates. It looks like the relationships is used for parents of a service?

Any insight you can give?

bzwei commented:
@kbrock Yes. `ServiceMixin` is included in both `Service` and `ServiceTemplate`. They both save resources in `service_resources` table. The template saves resources when a new template is defined. Service copies the resources and stored in the same table when it is ordered from the template. When you extract the parent resource to ancestry for service, I believe you need to do the same to the template. You'll need to test whether the ancestry relationship of a template is duplicated to the service when it is ordered.

bzwei commented:
@gmcculloug please comment.

kbrock commented:
@bzwei Where does `service_template` store the parent relationship?

`service.service_id` points to the parent service.

Does `service_template.service_template` point to the parent service template (that will later be transformed into a `service` reference in the created service?

gmcculloug commented:
@kbrock The service_templates are connected through the service_resource model.  A service_template bundle would have a service_resource for each of the service_templates it includes as a child.

kbrock commented:
@gmcculloug so while it may make sense for `service` to leverage ancestry, service_templates can not leverage.

The anti circular reference code is expensive here. as is looking up the tree.

kbrock commented:
fixed cops

kbrock commented:
ping @Fryguy this modifies a table schema

Fryguy commented on app/models/mixins/service_mixin.rb:
```diff
@@ -18,29 +18,33 @@ def add_resource(rsc, options = {})
     rsc_type = rsc.class.base_class.name.tableize
     raise _("Cannot connect service with nil ID.") if rsc.id.nil? && rsc_type == "service_templates"
 
+    # what is this?
```
These comments aren't very helpful for future readers. What about this do you not understand? Can you give more detail or at least ping someone for info?

Fryguy commented on app/models/mixins/service_mixin.rb:
```diff
@@ -104,23 +108,18 @@ def next_group_index(current_idx, direction = 1)
+    if child_svc.kind_of?(Service)
+      ancestor_ids.include?(child_svc.id)
+    elsif child_svc.kind_of?(ServiceTemplate)
+      !!circular_reference_check(child_svc)
```
@gmcculloug Can you review this bit?

kbrock replied:
I had deleted everything here. since ancestry takes care of services.
But for service templates, that was not possible. And possibly when it crosses those boundaries from one to the other.

kbrock replied:
this was renamed from `is_circular_reference?` to make rubo cop happy

Please also verify the `circular_reference_check` (line 123) w/ the change to the `kind_of?(Service)`

Fryguy commented on app/models/service.rb:
```diff
@@ -1,10 +1,12 @@
 class Service < ApplicationRecord
   DEFAULT_PROCESS_DELAY_BETWEEN_GROUPS = 120
 
+  has_ancestry :orphan_strategy => :destroy
```
@gmcculloug Can you verify the orphan_strategy here?  Seems ok to me, just want to verify that was your intent.

gmcculloug replied:
👍 Looks good.  The top-level service is the only one that is visible so the sub-services need to be removed when it is deleted.

Fryguy commented on app/models/service.rb:
```diff
@@ -69,55 +70,28 @@ def request_type
-  end
+  alias root_service root
+  alias services children
+  alias direct_service_children children
```
:heart_eyes:  Nice

Fryguy commented on app/models/service.rb:
```diff
@@ -69,55 +70,28 @@ def request_type
-    all_service_children + [self]
+    descendants(:from_depth => 2)
   end
+  Vmdb::Deprecation.deprecate_methods(self, :indirect_service_children)
```
Why deprecate?

kbrock replied:
most (all?) of the code that calls indirect is doing this to get the `subtree`.
kept it around because it was exposed to automate

Fryguy commented on app/models/service.rb:
```diff
@@ -69,55 +70,28 @@ def request_type
-    all_service_children.collect(&:direct_vms).flatten.compact
+    MiqPreloader.preload_and_map(indirect_service_children, :direct_vms)
   end
+  Vmdb::Deprecation.deprecate_methods(self, :indirect_vms)
```
Also, why deprecate?

Fryguy commented on app/models/service.rb:
```diff
@@ -69,55 +70,28 @@ def request_type
 
   def direct_vms
-    service_resources.collect { |sr| sr.resource.kind_of?(VmOrTemplate) ? sr.resource : nil }.flatten.compact
+    service_resources.where(:resource_type => 'VmOrTemplate').collect(&:resource)
```
Is this an N+1, or does service_resources already do the `includes`.

kbrock replied:
this is still an n+1, but it only does it for the vms (vs before it did it for every resource that existed)

Fryguy commented on app/models/service.rb:
```diff
@@ -69,55 +70,28 @@ def request_type
 
   def all_vms
-    direct_vms + indirect_vms
+    MiqPreloader.preload_and_map(subtree, :direct_vms)
```
This feels wrong...the old code did direct_vms + indirect_vms, but this does direct_vms, but on the entire subtree.  Is that to just filter out the VmOrTemplate bits?  Feels like a weird method to call on each node of the subtree.

kbrock replied:
```ruby
direct_vms = self.vms
indirect_vms = children.flat_map(&:vms)

all_vms = direct_vms + indirect_vms
all_vms = subtree.flat_map(&:vms)
```

is there a different way to get the vms for all nodes in the subtree?
I guess we could do something like `Vms.where(:service_id => subtree.select(:id))` - but the relations makes that tricky.

Fryguy replied:
I'll have to re-review my comment here...I think I might be wrong, but I need to double check

Fryguy replied:
Nevermind I'm wrong...I had it in my head that the ancestry tree included the VMs as well but it doesn't.

Fryguy commented on app/presenters/tree_builder_services.rb:
```diff
@@ -18,7 +18,7 @@ def set_locals_for_render
   # Get root nodes count/array for explorer tree
   def x_get_tree_roots(count_only, _options)
-    count_only_or_objects(count_only, rbac_filtered_objects(Service.where(:service_id => nil)), "name")
+    count_only_or_objects(count_only, rbac_filtered_objects(Service.where(:ancestry => nil)), "name")
```
Can you give Service.roots instead of the `ancestry => nil` bit?

kbrock replied:
+1. this blew up in the migrations. but here that would be much better - thanks

Fryguy commented on db/migrate/20160310170333_add_service_ancestry.rb:
```diff
@@ -0,0 +1,43 @@
+    add_index :services, :ancestry
+
+    Service.connection.schema_cache.clear!
+    Service.reset_column_information
```
This should not be necessary.  If it is, I'm very surprised and it's something that @lpichler should put on his TODO list.

Fryguy commented on spec/migrations/20160310170333_add_service_ancestry_spec.rb:
```diff
@@ -0,0 +1,50 @@
+
+  migration_context :up do
+    it "updates tree" do
+      s21  = service_stub.create!
```
Can you draw like an ASCII tree comment here so it's clear which s relates to which other s?

kbrock commented on app/models/service_template.rb:
```diff
@@ -33,6 +33,25 @@ class ServiceTemplate < ApplicationRecord
+
+  def descendants
+    klass = self.class
+    service_resources.flat_map do |s|
```
maybe try:

```ruby
service_resources.where(:resource_type => base_class.name).map { |s| s.resource }.flat_map { |r| r.descendants + [r] }
```

this will not use cache. scope instead? Not sure how many times traversed. (think `service_resources` are accessed again soon. so best to just load them - verify)

kbrock replied:
I'm just going to leave as the current code implemented it

kbrock replied:
```ruby
children.flat_map { |child| [child] + child.children }
```

kbrock commented:
rebased and fixed remaining cops.

@Fryguy / @gmcculloug do either of you need a call to help move this along? This fixes a serious performance issue.

Fryguy commented:
LGTM :+1: I just want confirmation from @gmcculloug

kbrock commented on app/models/service.rb:
```diff
@@ -50,16 +52,15 @@ def add_resource(rsc, options = {})
     super
   end
-  alias_method :<<, :add_resource
+  alias :<< add_resource
```
is this still needed

kbrock commented on app/models/service.rb:
```diff
@@ -1,10 +1,12 @@
   belongs_to :tenant
   belongs_to :service_template               # Template this service was cloned from
-  belongs_to :service                        # Parent Service
-  has_many :services, :dependent => :destroy # Child services
```
add `service_id` virtual column / alias

kbrock replied:
done

kbrock commented on db/migrate/20160310170333_add_service_ancestry.rb:
```diff
@@ -0,0 +1,37 @@
+
+  def update_service_parent(parent)
+    Service.where(:service_id => parent.try(:id)).each do |svc|
+      ancestry = [parent.try(:ancestry), parent.try(:id)].compact.join("/")
```
move this out

kbrock commented on app/models/service_template.rb:
```diff
@@ -33,6 +33,25 @@ class ServiceTemplate < ApplicationRecord
   virtual_has_one :custom_action_buttons, :class_name => "Array"
 
+  # almost feels like service_templates
+  def children
```
```ruby
def children
  service_templates
end
```

kbrock commented on db/migrate/20160310170333_add_service_ancestry.rb:
```diff
@@ -0,0 +1,37 @@
+
+    say_with_time("Converting Services from ancestry to service_id") do
+      Service.all.each do |service|
+        service.update_attributes(:service_id => service.ancestry.split("/").last.try(:to_i)) if service.ancestry
```
can `ancestry` be ""?
(answer: no, but just to be sure?)

gmcculloug commented on app/models/service.rb:
```diff
@@ -50,16 +52,20 @@ def add_resource(rsc, options = {})
+  def service_id
+    parent_id
   end
+  virtual_attribute :service_id, :integer
```
👍 Added for compatible with automate/reporting which had access to this column.

gmcculloug commented:
@kbrock Thanks for the updates.  Will merge when green.
